packages/go_router/lib/src/route.dart

```diff
@@ -1157,6 +1241,9 @@ class StatefulNavigationShellState extends State<StatefulNavigationShell>
   RouteMatchList? _matchListForBranch(int index) =>
       _branchLocations[route.branches[index]]?.value;
```

Comment:
Can you define a class for both branch and Navigator and use one list to store them, so that you don't need to keep two lists in sync?

Reply:
Also I think the current API is slightly weird. The navigator will become SizedBox.srink() when they are not loaded but still passed into `containerBuilder`, it seems counter intuitive if the `containerBuilder` decide to put some child that are never visited before or preloaded before onto the screen. We should probably have a better api.

Another thing is that we rely on both the _preloadNavigator to be build and the `containerBuilder` to put it as an offstage child to build the widget subtree. What if they use custom container builder and only builds the child at the index but throw the rest away?

I think the `containerBuilder` should have some kind of API that asks child for a given index, and if they do ask for it, we build the complete widget subtree regardless it is preloaded or not.

If branch is set to preload or has been built before but is not built by the containerBuilder, we can build a stack internally by `StatefulNavigationShell` and put them offstage

> but is not built by the containerBuilder

This will be the hardest part of the implementation, because `containerBuilder` can ask for widget build still throw it away. I think you may be able to build some special widget that wraps the navigator and only build subtree when it has not been built before. Not sure if that is a good idea though. Alternative, We could also document that if you ask for widget in `containerBuilder` you must also build it in the subtree.

I am curious if there is a better way.